yarn-project/bb-prover/src/prover/bb_prover_cache.ts

```diff
@@ -0,0 +1,71 @@
+import { TxExecutionRequest } from "@aztec/circuit-types";
+import { Fr, PrivateKernelTailCircuitPublicInputs, RecursiveProof, VerificationKeyAsFields, VerificationKeyData } from "@aztec/circuits.js";
+import { CacheableExecutionResult } from "@aztec/simulator";
+
+import * as crypto from "crypto";
+import * as fs from "fs/promises";
+import * as os from "os";
+import path from "path";
+
+import { ClassConverter, type ClassConverterInput } from "../../../foundation/src/json-rpc/class_converter.js";
+import { convertFromJsonBuffer, convertToJsonBuffer } from "../../../foundation/src/json-rpc/convert.js";
+
+const SHOULD_USE_PROVER_CACHE = !!process.env.PROVER_CACHE || !!process.env.PROVER_CACHE_DIR;
+const PROVER_CACHE_DIR = process.env.PROVER_CACHE_DIR || path.join(os.homedir(), '.aztec', 'proof-cache');
+
+export const PROVER_CLASS_CONVERTER_INPUT = {
+  fromString: {
+    Fr,
+  },
+  fromJSON: {},
+  fromBuffer: {
+    // For simulate
+    TxExecutionRequest,
+    CacheableExecutionResult,
+    // For bb app createProof
+    RecursiveProof,
+    PrivateKernelTailCircuitPublicInputs,
+    VerificationKeyAsFields,
+    // For bb rollup createProof
+    VerificationKeyData,
+  },
+};
+
+/**
+ * Create a result that serializes into proof-cache.
+ * Easy to wrap around any result, possibly needing to add classes to PROVER_CLASS_CONVERTER_INPUT as errors are hit.
+ * HOWEVER, note that wrapping interactions that cause side effects, e.g. creating files, will not have those side effects appear on cache load!
+ */
+export async function withProverCache<Result>(
+  label: string,
+  proofInputs: any[],
```

Comment:
Definitely hear you. It would be a bit tedious but I could fill in the needed serialization. ExecutionResult benefitted a good amount from serialization methods. 
I could be convinced for sure to use Bufferable instead, it was just a convenient catchall as I didn't want the cache feature causing serialization code burden slowing down development, while this can follow a mixed object structure.